### PR TITLE
Fixed CC generation

### PR DESCRIPTION
**Pull request by pavgra**
fixes https://github.com/OHDSI/WebAPI/issues/836

### CODE REVIEW

**Comment by anthonysena** [comment]
@pavgra still facing an issue with this CC: [cc_celecoxib.txt](https://github.com/OHDSI/WebAPI/files/2882044/cc_celecoxib.txt) while running on PDW but suspect you'll get the same result against SQL Server. Here is the stack trace:

```
2019-02-19 16:19:32.775 ERROR taskExecutor-5 org.springframework.batch.core.step.AbstractStep -  - Encountered an error executing step generateCohortCharacterization.generate in job generateCohortCharacterization
java.lang.RuntimeException: SQL contains more than one statement: IF XACT_STATE() = 1 COMMIT; CREATE TABLE  #Codesets (
  codeset_id int NOT NULL,
  concept_id bigint NOT NULL
)

	at org.ohdsi.sql.SqlTranslate.translateSingleStatementSqlWithPath(SqlTranslate.java:378)
	at org.ohdsi.sql.SqlTranslate.translateSingleStatementSql(SqlTranslate.java:351)
	at org.ohdsi.webapi.util.PreparedStatementRenderer.getSql(PreparedStatementRenderer.java:283)
	at org.ohdsi.webapi.util.PreparedStatementRendererCreator.getSql(PreparedStatementRendererCreator.java:35)
	at org.ohdsi.webapi.util.PreparedStatementUtils.collapsePreparedStatementCreators(PreparedStatementUtils.java:33)
	at org.ohdsi.webapi.util.PreparedStatementUtils.addAll(PreparedStatementUtils.java:60)
	at org.ohdsi.webapi.cohortcharacterization.GenerateCohortCharacterizationTasklet$CcTask.getSqlQueriesToRun(GenerateCohortCharacterizationTasklet.java:498)
	at org.ohdsi.webapi.cohortcharacterization.GenerateCohortCharacterizationTasklet$CcTask.getAnalysisQueriesOnCohort(GenerateCohortCharacterizationTasklet.java:171)
	at org.ohdsi.webapi.cohortcharacterization.GenerateCohortCharacterizationTasklet$CcTask.lambda$run$0(GenerateCohortCharacterizationTasklet.java:163)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.ohdsi.webapi.cohortcharacterization.GenerateCohortCharacterizationTasklet$CcTask.run(GenerateCohortCharacterizationTasklet.java:165)
	at org.ohdsi.webapi.cohortcharacterization.GenerateCohortCharacterizationTasklet$CcTask.access$000(GenerateCohortCharacterizationTasklet.java:126)
	at org.ohdsi.webapi.cohortcharacterization.GenerateCohortCharacterizationTasklet.prepareStatementCreators(GenerateCohortCharacterizationTasklet.java:116)
	at org.ohdsi.webapi.common.generation.CancelableTasklet.doTask(CancelableTasklet.java:37)
	at org.ohdsi.webapi.common.generation.CancelableTasklet.doTask(CancelableTasklet.java:17)
	at org.ohdsi.webapi.common.generation.TransactionalTasklet.lambda$execute$0(TransactionalTasklet.java:37)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at org.ohdsi.webapi.common.generation.TransactionalTasklet.execute(TransactionalTasklet.java:37)
	at org.springframework.batch.core.step.tasklet.TaskletStep$ChunkTransactionCallback.doInTransaction(TaskletStep.java:406)
	at org.springframework.batch.core.step.tasklet.TaskletStep$ChunkTransactionCallback.doInTransaction(TaskletStep.java:330)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at org.springframework.batch.core.step.tasklet.TaskletStep$2.doInChunkContext(TaskletStep.java:272)
	at org.springframework.batch.core.scope.context.StepContextRepeatCallback.doInIteration(StepContextRepeatCallback.java:81)
	at org.springframework.batch.repeat.support.RepeatTemplate.getNextResult(RepeatTemplate.java:374)
	at org.springframework.batch.repeat.support.RepeatTemplate.executeInternal(RepeatTemplate.java:215)
	at org.springframework.batch.repeat.support.RepeatTemplate.iterate(RepeatTemplate.java:144)
	at org.springframework.batch.core.step.tasklet.TaskletStep.doExecute(TaskletStep.java:257)
	at org.springframework.batch.core.step.AbstractStep.execute(AbstractStep.java:200)
	at org.springframework.batch.core.job.SimpleStepHandler.handleStep(SimpleStepHandler.java:148)
	at org.springframework.batch.core.job.AbstractJob.handleStep(AbstractJob.java:392)
	at org.springframework.batch.core.job.SimpleJob.doExecute(SimpleJob.java:135)
	at org.springframework.batch.core.job.AbstractJob.execute(AbstractJob.java:306)
	at org.springframework.batch.core.launch.support.SimpleJobLauncher$1.run(SimpleJobLauncher.java:135)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

The calling code is here:

https://github.com/OHDSI/WebAPI/blob/master/src/main/java/org/ohdsi/webapi/util/PreparedStatementUtils.java#L33

Since the CC drops in 211 statements, it was hard to find the offending caller but hoping these details help hone in on the root cause.

**Comment by anthonysena** [comment]
Thanks @pavgra for the update - that allowed me to move past the error above and successfully generate the cohort characterization. I did one more test, this time including strata in the characterization and ran into a different issue:

```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: org.springframework.jdbc.UncategorizedSQLException: ConnectionCallback; uncategorized SQLException for SQL []; SQL state [S0001]; error code [2714]; There is already an object named '#Codesets' in the database.; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: There is already an object named '#Codesets' in the database.
	at org.ohdsi.webapi.common.generation.StoppableTransactionalTasklet.waitForFuture(StoppableTransactionalTasklet.java:39)
	at org.ohdsi.webapi.common.generation.CancelableTasklet.doTask(CancelableTasklet.java:47)
	at org.ohdsi.webapi.common.generation.CancelableTasklet.doTask(CancelableTasklet.java:17)
	at org.ohdsi.webapi.common.generation.TransactionalTasklet.lambda$execute$0(TransactionalTasklet.java:37)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at org.ohdsi.webapi.common.generation.TransactionalTasklet.execute(TransactionalTasklet.java:37)
	at org.springframework.batch.core.step.tasklet.TaskletStep$ChunkTransactionCallback.doInTransaction(TaskletStep.java:406)
	at org.springframework.batch.core.step.tasklet.TaskletStep$ChunkTransactionCallback.doInTransaction(TaskletStep.java:330)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at org.springframework.batch.core.step.tasklet.TaskletStep$2.doInChunkContext(TaskletStep.java:272)
	at org.springframework.batch.core.scope.context.StepContextRepeatCallback.doInIteration(StepContextRepeatCallback.java:81)
	at org.springframework.batch.repeat.support.RepeatTemplate.getNextResult(RepeatTemplate.java:374)
	at org.springframework.batch.repeat.support.RepeatTemplate.executeInternal(RepeatTemplate.java:215)
	at org.springframework.batch.repeat.support.RepeatTemplate.iterate(RepeatTemplate.java:144)
	at org.springframework.batch.core.step.tasklet.TaskletStep.doExecute(TaskletStep.java:257)
	at org.springframework.batch.core.step.AbstractStep.execute(AbstractStep.java:200)
	at org.springframework.batch.core.job.SimpleStepHandler.handleStep(SimpleStepHandler.java:148)
	at org.springframework.batch.core.job.AbstractJob.handleStep(AbstractJob.java:392)
	at org.springframework.batch.core.job.SimpleJob.doExecute(SimpleJob.java:135)
	at org.springframework.batch.core.job.AbstractJob.execute(AbstractJob.java:306)
	at org.springframework.batch.core.launch.support.SimpleJobLauncher$1.run(
```

Here is the updated cc: 
[cc_celecoxib_with_stratum.txt](https://github.com/OHDSI/WebAPI/files/2882233/cc_celecoxib_with_stratum.txt)

**Comment by chrisknoll** [comment]
Error during execution:
```
 SQL state [S0001]; error code [104365]; Cannot create a global temporary table.; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: Cannot create a global temporary table.
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:84)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:360)
	at org.ohdsi.webapi.util.CancelableJdbcTemplate.batchUpdate(CancelableJdbcTemplate.java:136)
	at org.ohdsi.webapi.common.generation.CancelableTasklet.lambda$doTask$1(CancelableTasklet.java:39)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 3 more
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: Cannot create a global temporary table.
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:254)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1608)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:578)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:508)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7240)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:2869)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:243)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:218)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.execute(SQLServerPreparedStatement.java:493)
	at org.ohdsi.webapi.util.CancelableJdbcTemplate$1BatchUpdateConnectionCallback.doInConnection(CancelableJdbcTemplate.java:113)
	at org.ohdsi.webapi.util.CancelableJdbcTemplate$1BatchUpdateConnectionCallback.doInConnection(CancelableJdbcTemplate.java:101)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:353)
	... 6 more
2019-02-21 17:39:43.031 INFO taskExecutor-6 org.springframework.batch.core.launch.support.SimpleJobLauncher -  - Job: [SimpleJob: [name=generateCohortCharacterization]] completed with the following parameters: [{jobName=Generating cohort characterization 5 : Optum (PDW - DEV) (OPTUM-PDW), cohort_characterization_id=5, source_id=2, jobAuthor=anonymous, sessionId=fyqnqk4a, target_table=temp_cohort_fyqnqk4a, time=1550788722217}] and the following status: [FAILED]
```

**Comment by pavgra** [comment]
The thing is that from one side: the temp tables live inside PreparedStatement, so cannot be shared between different statements. On the other side, if I put two two distribution analyses - "Demographics Prior Observation Time" and "Demographics Time In Cohort" - into a single PreparedStatement, the statement fails as both analyses create e.g. #dem_time_data and then drop it, but the executor doesn't care about order, detects that the object already exists and fails. So, I see two ways: either precisely group SQLs into statements so that temp tables don't re-used in a single statement, or use ‘permenant temporary tables’. The second option looks better for me since such SQL could be exported and executed manually, which is not the case with the first option.

**Comment by pavgra** [comment]
@chrisknoll / @anthonysena , could you give it one more try?

**Comment by anthonysena** [comment]
Sure I'll give this a test.

**Comment by chrisknoll** [comment]
Still having an issue:

```
Caused by: java.util.concurrent.ExecutionException: org.springframework.jdbc.UncategorizedSQLException: ConnectionCallback; uncategorized SQLException for SQL [CREATE TABLE  [DEV_CDM_Optum_Extended_SES_v515].cknoll1.zne2g4lv_cov_ref   (covariate_id BIGINT,
	covariate_name VARCHAR(512),
	 analysis_id INT,
	analysis_name VARCHAR(512),
	concept_id INT
	)
WITH (LOCATION = USER_DB, DISTRIBUTION = HASH(analysis_id))]; SQL state [S0001]; error code [104370]; Cannot use the LOCATION option with a user table.; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: Cannot use the LOCATION option with a user table.

```

The problem here is that SqlRender sees the temp table (prepended with a `#`)  when it translates and adds the LOCATION=USER_DB for PDW, which isn't want we want in the case of the 'permanent temp table'. I'd suggest we take places that we want to use actual 'permanent temp tables' and take out the `#` from the table name, and replace it with a replaceable token that will make it unique to the session...I believe this is how the 'permanent temp tables' in cohort generation works.

File: src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java [hunk 1]
```diff
@@ -424,6 +425,31 @@ private String getStrataCohortTable(CcStrataEntity strata) {
...
+             *
+             * Therefore, there are two ways:
+             * - either precisely group SQLs into statements so that temp tables aren't re-used in a single statement,
+             * - or use ‘permenant temporary tables’
+             *
+             * The second option looks better since such SQL could be exported and executed manually,
+             * which is not the case with the first option.
+             */
+            if (ImmutableList.of(DBMSType.MS_SQL_SERVER.getOhdsiDB(), DBMSType.PDW.getOhdsiDB()).contains(source.getSourceDialect())) {
+              translatedSql = translatedSql
```
Comment (chrisknoll) [inline src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java]: The other way of handling this is handling the temp table switch prior to translation, which would probably make sense that all implementation would use the same 'temp table' strategy.

**Comment by chrisknoll** [comment]
@pavgra , do you think it is necessary to used prepared statements for the analysis queries? That's the core of the difficulty in using temp tables on sql server.

**Comment by pavgra** [comment]
@chrisknoll , let's try the approach with prepared statements once more. If the commited version won't work either, I'll switch to plain SQLs execution

**Comment by chrisknoll** [comment]
Ok, checked with and without strata and with the case where the cohorts are mutually exclusive. Looks good.

**Pull request merged**